// phonydata/lib/basic/locale.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.locale = void 0;
const __1 = require("..");
// Placeholders that should get overridden by locale-specific extensions.
function locale() {
    // Language/location specific
    (0, __1.define)('alphaNumericLower', 'abcdefghijklmnopqrstuvwxyz0123456789'.split(''));
    (0, __1.define)('alphaNumericUpper', 'ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'.split(''));
    (0, __1.define)('letterLower', 'abcdefghijklmnopqrstuvwxyz'.split(''));
    (0, __1.define)('letterUpper', 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split(''));
    (0, __1.define)('personName', function () {
        return this.givenName + ' ' + this.surname;
    });
    (0, __1.define)('sentence', function () {
        return this.loremSentence;
    });
    (0, __1.define)('sentencePunctuation', '..........?!'.split(''));
    (0, __1.define)('title', function () {
        return this.loremTitle;
    });
    (0, __1.define)('titleWords', function (num) {
        return this.loremTitleWords(num);
    });
    (0, __1.define)('word', function () {
        return this.loremWord;
    });
    (0, __1.define)('words', function (num) {
        return this.loremWords(num);
    });
    // Country specific
    (0, __1.define)('addressLine1', function () {
        return this.locality.addressLine1;
    });
    (0, __1.define)('buildingNumber', function () {
        return this.integer(1, 1000);
    });
    (0, __1.define)('city', function () {
        return this.locality.city;
    });
    (0, __1.define)('currencyValue', function () {
        return this.integer(0, 10000) / 100;
    });
    (0, __1.define)('givenName', function () {
        return this.random < 0.5 ? this.givenNameFemale : this.givenNameMale;
    });
    (0, __1.define)('givenNameFemale', function () {
        return this.loremTitleWords(1);
    });
    (0, __1.define)('givenNameMale', function () {
        return this.loremTitleWords(1);
    });
    (0, __1.define)('locality', function () {
        const state = this.loremTitleWords(1);
        const stateCode = (state + 'XXX').substring(0, 3).toUpperCase();
        return {
            addressLine1: this.buildingNumber + ' ' + this.streetName,
            city: this.loremTitleWords(1),
            stateOrProvince: state,
            stateOrProvinceCode: stateCode,
            postCode: this.format('ZZZ ZZZ'),
        };
    });
    (0, __1.define)('phoneNumber', function () {
        return this.integer(2, 9).toString() + this.format('##-###-####');
    });
    (0, __1.define)('postCode', function () {
        return this.locality.postCode;
    });
    (0, __1.define)('stateOrProvince', function () {
        return this.locality.stateOrProvince;
    });
    (0, __1.define)('stateOrProvinceCode', function () {
        return this.locality.stateOrProvinceCode;
    });
    (0, __1.define)('streetName', function () {
        return this.loremTitleWords(2);
    });
    (0, __1.define)('surname', function () {
        return this.loremTitleWords(1);
    });
}
exports.locale = locale;
